packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -398,15 +469,40 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
 
     assert(maxLength > 0);
 
-    // If already at the maximum and tried to enter even more, keep the old
-    // value.
-    if (oldValue.text.characters.length == maxLength && !oldValue.composing.isValid) {
-      return oldValue;
+    switch (maxLengthEnforcement) {
+      case MaxLengthEnforcement.none:
+        return newValue;
+      case MaxLengthEnforcement.enforced:
+        // If already at the maximum and tried to enter even more, keep the old
+        // value.
+        if (oldValue.text.characters.length == maxLength) {
+          return oldValue;
+        }
+
+        // Enforced to return a truncated value.
+        return truncate(
+          newValue,
+          maxLength,
+          keepComposingRange: oldValue.composing.isValid,
```

Comment:
I think `keepComposingRange` should always be false for `.enforced`?

Reply:
maxLength = 5

1. `TextEditingValue(text: 'abcd', composing: TextRange(2, 4))`
2. Input "e", `TextEditingValue(text: 'abcde', composing: TextRange(2, 5))`
3. Input "f", `TextEditingValue(text: 'abcde', composing: TextRange(2, 5))` <- At here truncate comes in, and it should keep the composing range.

Reply:
I'm kinda forget what this flag for...It seems we can handle it directly in the truncate method. I'll double check that.

Reply:
Yep it's useless. Removing